src/main/native/org_cryptomator_windows_keychain_WindowsHello_Native.cpp

```diff
@@ -126,37 +148,72 @@ IBuffer DeriveKeyUsingHKDF(const IBuffer& inputData, const IBuffer& salt, uint32
 
     try {
-        // The first time this is used a key-pair will be generated using the common name
-        auto result = KeyCredentialManager::RequestCreateAsync(keyId,
-            KeyCredentialCreationOption::FailIfExists).get();
-
-        if (result.Status() == KeyCredentialStatus::CredentialAlreadyExists) {
-            result = KeyCredentialManager::OpenAsync(keyId).get();
-        }
-        else if (result.Status() != KeyCredentialStatus::Success) {
-            std::cerr << "Failed to retrieve Windows Hello credential." << std::endl;
-            return false;
+        std::vector<uint8_t> signatureData;
+        bool foundInCache = false;
+
+        {
+            // Lock for thread safety
+            std::lock_guard<std::mutex> lock(cacheMutex);
+            auto it = keyCache.find(keyId);
+            if (it != keyCache.end()) {
+                signatureData = it->second;
+                if (!UnprotectMemory(signatureData)) {
+                    throw std::runtime_error("Failed to unprotect memory.");
+                }
+                foundInCache = true;
+            }
         }
 
-        const auto signature = result.Credential().RequestSignAsync(challengeBuffer).get();
-        if (signature.Status() != KeyCredentialStatus::Success) {
-            if (signature.Status() != KeyCredentialStatus::UserCanceled) {
-                std::cerr << "Failed to sign challenge using Windows Hello." << std::endl;
+        if (!foundInCache) {
+            auto result = KeyCredentialManager::RequestCreateAsync(keyId, KeyCredentialCreationOption::FailIfExists).get();
+
+            if (result.Status() == KeyCredentialStatus::CredentialAlreadyExists) {
+                result = KeyCredentialManager::OpenAsync(keyId).get();
             }
-            return false;
+            else if (result.Status() != KeyCredentialStatus::Success) {
+                std::cerr << "Failed to retrieve Windows Hello credential." << std::endl;
+                return false;
+            }
+
+            const auto signature = result.Credential().RequestSignAsync(challengeBuffer).get();
+
+            if (signature.Status() != KeyCredentialStatus::Success) {
+                if (signature.Status() != KeyCredentialStatus::UserCanceled) {
+                    std::cerr << "Failed to sign challenge using Windows Hello." << std::endl;
+                }
+                return false;
+            }
+
+            signatureData = iBufferToVector(signature.Result());
+            std::vector<uint8_t> protectedCopy = signatureData;
+
+            if (!ProtectMemory(protectedCopy)) {
+                throw std::runtime_error("Failed to protect memory.");
+            }
+
+            // Store in cache
+            {
+                std::lock_guard<std::mutex> lock(cacheMutex);
+                keyCache[keyId] = protectedCopy;
+            }
+            std::fill(protectedCopy.begin(), protectedCopy.end(), 0);
```

Comment:
Move this to its own method

Reply:
@infeo I implemented the changes:
- CRYPTPROTECTMEMORY_BLOCK_SIZE
- Move this to its own method

I could not spot any places, where zeroing vectors has been forgotten.

On testing, I noticed a missing piece on the cache implementation: when a signature is cached and a second vault created afterwards (with the cached signature), HMAC verification failed for the second vault, when you end the app, start it again and try to unlock the second vault first. This was due to a newly created challenge for the second vault.

This is fixed now too by using a fixed challenge in the `WindowsHello` class, as discussed for this last part of the cache implementation.